### PR TITLE
Add mixed precision training support

### PR DESCRIPTION
**Pull request by workofart**
## Description
This is a very rough attempt in mixed precision training.

Essentially, in the GPT-2 example change in this PR, we are defining these dtypes:
- model params: bfloat16 (this is hard-coded for certain bf16-compatible backends for now)
- optimizer states:  float32 (Adam casts bf16 grads to fp32 before updating momentum estimates)
- gradient accumulation: float32
- LayerNorm stats: float32
- Cross entropy logits: float32  
- activations:  bfloat16 (after going through LayerNorm/attention/feedforward)

## Tests

| Metric | `main` branch fp32 | Mixed bf16 | Delta |
|---|---|---|---|
| steps/s | 3.764 | 3.706 | -1.54% |
| tokens/s | 7708 | 7591 | -1.52% |
| active memory | 596.69 MiB | 497.25 MiB | -16.67% |
| peak memory | 7.07 GiB | 6.99 GiB | -1.13% |

### CODE REVIEW

File: examples/gpt_2.py [hunk 1]
```diff
@@ -259,6 +275,7 @@ def sample(self, model: GPT2, input_ids: Array) -> Tensor:
             "max_seq_len": 1024,  # GPT-2 uses 1024
             "num_decoder_layers": 12,  # GPT-2 uses 12
             "activation_checkpointing": True,
+            "parameter_dtype": "bfloat16",
```
Comment (workofart) [inline examples/gpt_2.py:278]: if a backend doesn't support bf16, this will crash. I'm too lazy to complicate/generalize this for now, since I'm training on MLX currently.
